gulp-load-plugins as a dependency of a dependency not loading plugins

EDIT: Remove all comments until this one: https://github.com/jackfranklin/gulp-load-plugins/issues/56#issuecomment-71681420

I *think* this is the same as #35.

I have the following project which I use to quickly load up gulp tasks in projects where I work without duplicating code: https://github.com/Lostmyname/lmn-gulp-tasks

Pretty much the same file structure described in [this comment](https://github.com/jackfranklin/gulp-load-plugins/issues/35#issuecomment-49170938), and it's looking in the top package.json, not the bottom one:

```
~/Sites/lmn/monkey master*  6m 44s  ⇡
❯ npm install lmn-gulp-tasks
npm WARN package.json lmn-monkey@3.0.1 No repository field.
npm WARN package.json browsernizr@1.0.2 No repository field.
lmn-gulp-tasks@1.0.0 node_modules/lmn-gulp-tasks
├── jshint-stylish@1.0.0 (log-symbols@1.0.1, text-table@0.2.0, string-length@1.0.0, chalk@0.5.1)
├── gulp-plumber@0.6.6 (through2@0.6.3)
├── gulp-jshint@1.9.0 (rcloader@0.1.2, minimatch@1.0.0, lodash@2.4.1, through2@0.6.3)
├── gulp-util@3.0.2 (array-differ@1.0.0, object-assign@2.0.0, beeper@1.0.0, array-uniq@1.0.2, lodash._reinterpolate@2.4.1, replace-ext@0.0.1, minimist@1.1.0, chalk@0.5.1, multipipe@0.1.2, vinyl@0.4.6, through2@0.6.3, lodash.template@2.4.1, dateformat@1.0.11)
├── gulp-minify-css@0.4.3 (memory-cache@0.0.5, bufferstreams@0.0.2, through2@0.6.3, vinyl-sourcemaps-apply@0.1.4, clean-css@3.0.7)
├── stylus@0.49.3 (css-parse@1.7.0, mkdirp@0.3.5, debug@2.1.1, glob@3.2.11, sax@0.5.8, source-map@0.1.43)
├── gulp-compass@2.0.3 (which@1.0.8, through2@0.6.3)
├── jshint@2.6.0 (strip-json-comments@1.0.2, underscore@1.6.0, console-browserify@1.1.0, exit@0.1.2, minimatch@1.0.0, shelljs@0.3.0, cli@0.6.5, htmlparser2@3.8.2)
├── gulp-jscs@1.4.0 (object-assign@2.0.0, through2@0.6.3, jscs@1.10.0)
├── gulp-autoprefixer@2.1.0 (object-assign@2.0.0, through2@0.6.3, vinyl-sourcemaps-apply@0.1.4, autoprefixer-core@5.0.0)
└── browserify@8.1.1 (https-browserify@0.0.0, tty-browserify@0.0.0, builtins@0.0.7, constants-browserify@0.0.1, process@0.10.0, path-browserify@0.0.0, os-browserify@0.1.2, isarray@0.0.1, inherits@2.0.1, commondir@0.0.1, string_decoder@0.10.31, stream-browserify@1.0.0, defined@0.0.0, domain-browser@1.1.3, shell-quote@0.0.1, xtend@3.0.0, shallow-copy@0.0.1, deep-equal@0.2.1, querystring-es3@0.2.1, assert@1.3.0, timers-browserify@1.2.0, punycode@1.2.4, util@0.10.3, vm-browserify@0.0.4, events@1.0.2, parents@1.0.1, console-browserify@1.1.0, subarg@1.0.0, url@0.10.2, readable-stream@1.0.33, http-browserify@1.7.0, deps-sort@1.3.5, through2@1.1.1, resolve@0.7.4, shasum@1.0.1, buffer@3.0.1, concat-stream@1.4.7, glob@4.3.5, duplexer2@0.0.2, browser-resolve@1.6.0, JSONStream@0.8.4, labeled-stream-splicer@1.0.2, browser-pack@3.2.0, crypto-browserify@3.9.10, umd@2.1.0, insert-module-globals@6.2.0, syntax-error@1.1.2, browserify-zlib@0.1.4, module-deps@3.6.4)

~/Sites/lmn/monkey master*  15s  ⇡
❯ gulp
[16:00:49] Using gulpfile ~/Sites/lmn/monkey/gulpfile.js
[16:00:49] Starting 'html'...
[16:00:49] Starting 'js-quality'...
[16:00:49] 'js-quality' errored after 8.71 ms
[16:00:49] Error: Cannot find module 'gulp-plumber'
    at Function.Module._resolveFilename (module.js:338:15)
    at Function.Module._load (module.js:280:25)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.defineProperty.get (/Users/callumacrae/Sites/lmn/monkey/node_modules/gulp-load-plugins/index.js:56:18)
    at Gulp.jsQualityTask (/Users/callumacrae/Sites/lmn/monkey/node_modules/lmn-gulp-tasks/tasks/js-quality.js:15:33)
    at module.exports (/Users/callumacrae/Sites/lmn/monkey/node_modules/gulp/node_modules/orchestrator/lib/runTask.js:34:7)
    at Gulp.Orchestrator._runTask (/Users/callumacrae/Sites/lmn/monkey/node_modules/gulp/node_modules/orchestrator/index.js:273:3)
    at Gulp.Orchestrator._runStep (/Users/callumacrae/Sites/lmn/monkey/node_modules/gulp/node_modules/orchestrator/index.js:214:10)
    at Gulp.Orchestrator.start (/Users/callumacrae/Sites/lmn/monkey/node_modules/gulp/node_modules/orchestrator/index.js:134:8)
[16:00:49] Finished 'html' after 62 ms
```

If I replace `plugins.plumber(` with just `plumber(` and require it manually, it errors about the next module instead.

Interestingly, it doesn't happen if I npm link the `lmn-gulp-tasks` project in—it only happens when I actually install it through npm.

Any ideas what's happening?

--- a/.editorconfig
+++ b/.editorconfig
@@ -1,0 +1,7 @@
+root = true
+
+[*]
+end_of_line = lf
+insert_final_newline = true
+indent_style = space
+indent_size = 2

--- a/.gitignore
+++ b/.gitignore
@@ -1,1 +1,1 @@
-node_modules
+/node_modules

--- a/index.js
+++ b/index.js
@@ -30,7 +30,8 @@ module.exports = function(options) {
     requireFn = options.requireFn;
   } else if(typeof config === 'string') {
     requireFn = function (name) {
-      return require(path.join(path.dirname(config), 'node_modules', name));
+      var searchFor = path.join('node_modules', name);
+      return require(findup(searchFor, {cwd: path.dirname(config)}));
     };
   } else {
     requireFn = require;

--- a/index.js
+++ b/index.js
@@ -15,6 +15,7 @@ function camelize(str) {
 
 module.exports = function(options) {
   var finalObject = {};
+  var configObject, requireFn;
   options = options || {};
 
   var pattern = arrayify(options.pattern || ['gulp-*', 'gulp.*']);
@@ -23,19 +24,26 @@ module.exports = function(options) {
   var replaceString = options.replaceString || /^gulp(-|\.)/;
   var camelizePluginName = options.camelize === false ? false : true;
   var lazy = 'lazy' in options ? !!options.lazy : true;
-  var requireFn = options.requireFn || require;
   var renameObj = options.rename || {};
 
-  if (typeof config === 'string') {
-    config = require(config);
+  if(typeof options.requireFn === 'function') {
+    requireFn = options.requireFn;
+  } else if(typeof config === 'string') {
+    requireFn = function (name) {
+      return require(path.join(path.dirname(config), 'node_modules', name));
+    };
+  } else {
+    requireFn = require;
   }
 
-  if(!config) {
+  configObject = (typeof config === 'string') ? require(config) : config;
+
+  if(!configObject) {
     throw new Error('Could not find dependencies. Do you have a package.json file in your project?');
   }
 
   var names = scope.reduce(function(result, prop) {
-    return result.concat(Object.keys(config[prop] || {}));
+    return result.concat(Object.keys(configObject[prop] || {}));
   }, []);
 
   pattern.push('!gulp-load-plugins');

--- a/index.js
+++ b/index.js
@@ -15,7 +15,8 @@ function camelize(str) {
 
 module.exports = function(options) {
   var finalObject = {};
-  var configObject, requireFn;
+  var configObject;
+  var requireFn;
   options = options || {};
 
   var pattern = arrayify(options.pattern || ['gulp-*', 'gulp.*']);
@@ -30,6 +31,8 @@ module.exports = function(options) {
     requireFn = options.requireFn;
   } else if(typeof config === 'string') {
     requireFn = function (name) {
+      // This searches up from the specified package.json file, making sure
+      // the config option behaves as expected. See issue #56.
       var searchFor = path.join('node_modules', name);
       return require(findup(searchFor, {cwd: path.dirname(config)}));
     };

--- a/package.json
+++ b/package.json
@@ -3,7 +3,7 @@
   "version": "0.8.0",
   "description": "Automatically load any gulp plugins in your package.json",
   "scripts": {
-    "test": "mocha test.js"
+    "test": "mocha"
   },
   "files": [
     "index.js"
@@ -28,6 +28,7 @@
   "license": "MIT",
   "dependencies": {
     "findup-sync": "^0.2.1",
+    "gulp-uglify": "^1.1.0",
     "multimatch": "2.0.0"
   },
   "devDependencies": {

--- a/package.json
+++ b/package.json
@@ -23,12 +23,12 @@
     "kombucha",
     "Nicolas Froidure",
     "Sindre Sorhus",
-    "Julien Fontanet <julien.fontanet@vates.fr>"
+    "Julien Fontanet <julien.fontanet@vates.fr>",
+    "Callum Macrae"
   ],
   "license": "MIT",
   "dependencies": {
     "findup-sync": "^0.2.1",
-    "gulp-uglify": "^1.1.0",
     "multimatch": "2.0.0"
   },
   "devDependencies": {

--- a/test/index.js
+++ b/test/index.js
@@ -13,7 +13,7 @@ var gulpLoadPlugins = (function() {
 
   var proxyquire = require('proxyquire').noCallThru();
 
-  return proxyquire('./', {
+  return proxyquire('../', {
     'gulp-foo': wrapInFunc({ name: 'foo' }),
     'gulp-bar': wrapInFunc({ name: 'bar' }),
     'gulp-foo-bar': wrapInFunc({ name: 'foo-bar' }),
@@ -185,5 +185,12 @@ describe('with lazy loading', function() {
   it('does when the property is accessed', function() {
     x.insert();
     assert(spy.called);
+  });
+});
+
+describe('requiring from another directory', function() {
+  it('allows you to use in a lower directory', function() {
+    var plugins = require('../')();
+    assert.ok(typeof plugins.test === 'function');
   });
 });

--- a/test/package.json
+++ b/test/package.json
@@ -1,0 +1,5 @@
+{
+  "dependencies": {
+    "gulp-test": "*"
+  }
+}

